Add raft log truncation queue.

BramGruneir opened the pull request:
Fixes #2877.

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/2997)

<!-- Reviewable:end -->

BramGruneir committed:
Add raft log truncation queue.

tamird commented on storage/client_raft_log_queue_test.go:
```diff
@@ -0,0 +1,94 @@
+		if err != nil {
+			return err
+		}
+		fmt.Printf("***** first:%d, current:%d\n", firstIndex, currentIndex)
```
did not review in detail, but this doesn't seem right. perhaps move this detail to the error you're returning just below?

BramGruneir replied:
ah, this was just leftover debug output, removed.

tamird commented on storage/client_raft_log_queue_test.go:
```diff
@@ -0,0 +1,94 @@
+		return nil
+	})
+
+	// Restart the store to tear down the test cleanly.
```
I see that this is copied from elsewhere, but it's troubling. @bdarnell is it still necessary?

bdarnell replied:
Even if it's still necessary in the other tests we shouldn't copy it here without good reason; I don't think this test would run into as many tricky edge cases as `replica_gc_queue_test`.  And it looks like it's no longer needed in that test either, so I'll send a separate PR to clean that up.

BramGruneir replied:
I've removed it.  Let's see if it's needed.

tamird commented on storage/raft_log_queue.go:
```diff
@@ -0,0 +1,127 @@
+		if log.V(1) {
+			log.Infof("truncating the raft log of range %d from %d to %d", r.Desc().RangeID, firstIndex, oldestIndex)
+		}
+		_, err = client.SendWrapped(r, r.context(), &roachpb.TruncateLogRequest{Index: oldestIndex})
```
nit: `err :=`

BramGruneir replied:
done

bdarnell commented on storage/client_raft_log_queue_test.go:
```diff
@@ -0,0 +1,93 @@
+		}
+	}
+
+	// Turn the raft log queue back on.
```
Stale comment

BramGruneir replied:
removed

bdarnell commented on storage/client_raft_log_queue_test.go:
```diff
@@ -0,0 +1,93 @@
+		}
+		return nil
+	})
+
```
We should also make sure that running `ForceRaftLogScan` a second time is a no-op (i.e. that the `TruncateLog` command itself didn't push us over the threshold).

BramGruneir replied:
Good call.  I'll add the post op check.

bdarnell commented on storage/raft_log_queue.go:
```diff
@@ -0,0 +1,127 @@
+
+	// Find the oldest index still in use by the range.
+	oldestIndex := raftStatus.Applied
+	for _, progress := range raftStatus.Progress {
```
The `Progress` and `Match` fields are only populated if `raftStatus.RaftState == raft.StateLeader`; we should abort if that's not true.

BramGruneir replied:
done, added a function to replica that checks if a replica is the raft leader and returns the raft status along with it.

bdarnell commented on storage/raft_log_queue.go:
```diff
@@ -0,0 +1,127 @@
+	if raftStatus == nil {
+		return 0, 0, util.Errorf("could not get raft status for %s", desc)
+	}
+	firstIndex, err := r.FirstIndex()
```
Move this to the beginning or end of the function instead of in the middle of the stuff using `raftStatus`. (or just call `FirstIndex` from `shouldQueue` and `process`; putting it in this function is awkward)

BramGruneir replied:
Sure, can do.  I just was trying to minimize the amount of repeated code is all.

BramGruneir replied:
fixed

bdarnell commented on storage/raft_log_queue.go:
```diff
@@ -0,0 +1,127 @@
+}
+
+func (*raftLogQueue) needsLeaderLease() bool {
+	return true
```
The distinction between the lease holder and raft leadership is subtle here. In order to see the log positions, we need raft leadership, which may not coincide with the lease. So I think `needsLeaderLease` should return false, and then `shouldQueue` should only return true for replicas that are the raft leader. (and we should figure out if there's some way to test this; I don't think we currently have any test cases where the raft leader is known to differ from the lease holder)

BramGruneir replied:
Ah, excellent point.  You're right, I had conflated the two concepts in this PR.  I'll make the necessary adjustments.

BramGruneir replied:
done.

bdarnell commented on storage/raft_log_queue.go:
```diff
@@ -0,0 +1,127 @@
+		if log.V(1) {
+			log.Infof("truncating the raft log of range %d from %d to %d", r.Desc().RangeID, firstIndex, oldestIndex)
+		}
+		_, err := client.SendWrapped(r, r.context(), &roachpb.TruncateLogRequest{Index: oldestIndex})
```
This needs to be sent through `rlq.db` instead of `r`, so the client layer can handle any `NotLeaderErrors` that are returned (because here we need the lease, not raft leadership)

BramGruneir replied:
Of course.  Done.

BramGruneir commented:
Review status: 0 of 4 files reviewed at latest revision, 9 unresolved discussions.

---

<sup>**[storage/client_raft_log_queue_test.go, line 84 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2997#-K2DgvkEW0Hq3tashnk1-r1-84)** ([raw file](https://github.com/cockroachdb/cockroach/blob/dce8bd4e5c804defcaa3ff1f8fabca07332f9171/storage/client_raft_log_queue_test.go#L84)):</sup>
Done.
<!-- Reviewable comment -K2NlgU9a8mwiPG7v6o1:-499146009 -->

---

<sup>**[storage/client_raft_log_queue_test.go, line 91 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2997#-K2DgvkEW0Hq3tashnk1-r1-91)** ([raw file](https://github.com/cockroachdb/cockroach/blob/dce8bd4e5c804defcaa3ff1f8fabca07332f9171/storage/client_raft_log_queue_test.go#L91)):</sup>
Done.
<!-- Reviewable comment -K2NlRqs1JfNzO0Qglft:-499146009 -->

---

<sup>**[storage/client_raft_log_queue_test.go, line 72 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2997#-K2DgvkEW0Hq3tashnk1-r2-72)** ([raw file](https://github.com/cockroachdb/cockroach/blob/9054685576ebb28a44c591469e69317b94e57089/storage/client_raft_log_queue_test.go#L72)):</sup>
Done.
<!-- Reviewable comment -K2NlTDBLngH0YFZNE5v:-499146009 -->

---

<sup>**[storage/client_raft_log_queue_test.go, line 89 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2997#-K2DgvkEW0Hq3tashnk1-r2-89)** ([raw file](https://github.com/cockroachdb/cockroach/blob/9054685576ebb28a44c591469e69317b94e57089/storage/client_raft_log_queue_test.go#L89)):</sup>
Done.
<!-- Reviewable comment -K2NlSN3_z3-p-fG10aF:-499146009 -->

---

<sup>**[storage/raft_log_queue.go, line 118 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2997#-K2DgvkEW0Hq3tashnk2-r1-118)** ([raw file](https://github.com/cockroachdb/cockroach/blob/dce8bd4e5c804defcaa3ff1f8fabca07332f9171/storage/raft_log_queue.go#L118)):</sup>
Done.
<!-- Reviewable comment -K2NlX9hAPZFksuTkCwa:-499146009 -->

---

<sup>**[storage/raft_log_queue.go, line 59 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2997#-K2DgvkEW0Hq3tashnk2-r2-59)** ([raw file](https://github.com/cockroachdb/cockroach/blob/9054685576ebb28a44c591469e69317b94e57089/storage/raft_log_queue.go#L59)):</sup>
Done.
<!-- Reviewable comment -K2NlQ5IMHPzduvxll5E:-499146009 -->

---

<sup>**[storage/raft_log_queue.go, line 74 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2997#-K2DgvkEW0Hq3tashnk2-r2-74)** ([raw file](https://github.com/cockroachdb/cockroach/blob/9054685576ebb28a44c591469e69317b94e57089/storage/raft_log_queue.go#L74)):</sup>
Done.
<!-- Reviewable comment -K2NlPOpdo50SDAOKgvq:-499146009 -->

---

<sup>**[storage/raft_log_queue.go, line 81 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2997#-K2DgvkEW0Hq3tashnk2-r2-81)** ([raw file](https://github.com/cockroachdb/cockroach/blob/9054685576ebb28a44c591469e69317b94e57089/storage/raft_log_queue.go#L81)):</sup>
Done.
<!-- Reviewable comment -K2NlVtf9P1unegY0vnq:-499146009 -->

---

<sup>**[storage/raft_log_queue.go, line 118 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2997#-K2DgvkEW0Hq3tashnk2-r2-118)** ([raw file](https://github.com/cockroachdb/cockroach/blob/9054685576ebb28a44c591469e69317b94e57089/storage/raft_log_queue.go#L118)):</sup>
Done.
<!-- Reviewable comment -K2NlXVMRLaum2BtZ_CD:-499146009 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/2997)
<!-- Sent from Reviewable.io -->

bdarnell commented:
LGTM
<!-- Reviewable comment -K2NvvqvtqRBP_O95-rG:676834233 -->

---

Review status: 0 of 4 files reviewed at latest revision, 8 unresolved discussions.

---

<sup>**[storage/client_raft_log_queue_test.go, line 31 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2997#-K2NvcMIYlB3lyCexipI)** ([raw file](https://github.com/cockroachdb/cockroach/blob/2f07fdb6342c98c291f08f2d2260b0772b7af258/storage/client_raft_log_queue_test.go#L31)):</sup>
This comment was incorrectly copied.

Add a TODO somewhere in this file to test a case where the raft leader and the lease holder differ.
<!-- Reviewable comment -K2NvcMIYlB3lyCexipJ:1016408783 -->

---

<sup>**[storage/raft_log_queue.go, line 87 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2997#-K2NtPC-p3PSnmT0EJLH)** ([raw file](https://github.com/cockroachdb/cockroach/blob/2f07fdb6342c98c291f08f2d2260b0772b7af258/storage/raft_log_queue.go#L87)):</sup>
Check for `err` before `raftLeader`. When `err` is non-nil other return values are generally considered undefined.
<!-- Reviewable comment -K2NtPC-p3PSnmT0EJLI:-1008274781 -->

---

<sup>**[storage/raft_log_queue.go, line 103 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2997#-K2NrX79KWxKj139_S_C)** ([raw file](https://github.com/cockroachdb/cockroach/blob/2f07fdb6342c98c291f08f2d2260b0772b7af258/storage/raft_log_queue.go#L103)):</sup>
Both these variables are unsigned, so if `oldestIndex < firstIndex` bad things could happen. This can happen if we have no information about a particular follower (`oldestIndex == 0`), or if a follower is so far behind that it needs to be caught up via snapshot.
<!-- Reviewable comment -K2NrX79KWxKj139_S_D:-2091516823 -->

---

<sup>**[storage/raft_log_queue.go, line 111 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2997#-K2NtZ6KajaoA7xpUJR0)** ([raw file](https://github.com/cockroachdb/cockroach/blob/2f07fdb6342c98c291f08f2d2260b0772b7af258/storage/raft_log_queue.go#L111)):</sup>
Ditto
<!-- Reviewable comment -K2NtZ6KajaoA7xpUJR1:-499318459 -->

---

<sup>**[storage/replica.go, line 1622 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2997#-K2Ntig-ZTJKZ0GYqcao)** ([raw file](https://github.com/cockroachdb/cockroach/blob/2f07fdb6342c98c291f08f2d2260b0772b7af258/storage/replica.go#L1622)):</sup>
It's good that this returns the  `raft.Status` used, since we have to use a consistent snapshot of the status when we're checking the `Progress`. But this pattern doesn't really scale - if we had multiple methods that followed this pattern they wouldn't be composable. It would be better to get the status from one method and then pass it into a second `isRaftLeader(*raft.Status)`.
<!-- Reviewable comment -K2Ntig-ZTJKZ0GYqcap:-1476071244 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/2997)
<!-- Sent from Reviewable.io -->

tschottdorf commented on storage/client_raft_log_queue_test.go:
```diff
@@ -0,0 +1,105 @@
+	"github.com/cockroachdb/cockroach/util/leaktest"
+)
+
+// TestReplicaGCQueueDropReplicaOnScan verifies that the range GC queue
```
comment is wrong.

tschottdorf commented on storage/client_raft_log_queue_test.go:
```diff
@@ -0,0 +1,105 @@
+	mtc := startMultiTestContext(t, 3)
+	defer mtc.Stop()
+
+	// Write a single values to ensure we have a leader.
```
s/values/value/

tschottdorf commented on storage/client_raft_log_queue_test.go:
```diff
@@ -0,0 +1,105 @@
+	}
+
+	// Write a collection of values to increase the raft log.
+	for i := 0; i < storage.RaftLogQueueLogSizeThreshold+1; i++ {
```
how many is that gonna write? Or, more importantly, how long does the test run?

tschottdorf commented on storage/client_raft_log_queue_test.go:
```diff
@@ -0,0 +1,105 @@
+	if err != nil {
+		t.Fatal(err)
+	}
+	if currentIndex != finalIndex {
```
is this stable? seems to me that the store could be using the range for something that requires raft (for example waking it up to renew lease for gossip).

tschottdorf commented on storage/raft_log_queue.go:
```diff
@@ -0,0 +1,141 @@
+	RaftLogQueueTimerDuration = time.Second
+	// RaftLogQueueLogSizeThreshold is the number of raft log entries after
+	// which truncation is required.
+	RaftLogQueueLogSizeThreshold = 1
```
oh, one - is that really a good default?

tschottdorf commented on storage/raft_log_queue.go:
```diff
@@ -0,0 +1,141 @@
+// raft status.
+func getOldestIndex(raftStatus raft.Status) uint64 {
+	// Find the oldest index still in use by the range.
+	oldestIndex := raftStatus.Applied
```
why isn't this initialized to `math.MaxInt64` (or whatever the type is)?

tschottdorf commented on storage/raft_log_queue.go:
```diff
@@ -0,0 +1,141 @@
+	priority float64) {
+	raftLeader, raftStatus, err := r.isRaftLeader()
+	if !raftLeader {
+		if err != nil {
```
so if `raftLeader` is true, you ignore the error?

tschottdorf commented on storage/raft_log_queue.go:
```diff
@@ -0,0 +1,141 @@
+	return truncatableIndexes > RaftLogQueueLogSizeThreshold, float64(truncatableIndexes)
+}
+
+// process synchronously invokes for each replica that requires a raft log
```
grammar

tschottdorf commented on storage/raft_log_queue.go:
```diff
@@ -0,0 +1,141 @@
+func (rlq *raftLogQueue) process(now roachpb.Timestamp, r *Replica, _ *config.SystemConfig) error {
+	raftLeader, raftStatus, err := r.isRaftLeader()
+	if !raftLeader {
+		if err != nil {
```
ditto

tschottdorf commented on storage/raft_log_queue.go:
```diff
@@ -0,0 +1,141 @@
+	oldestIndex := getOldestIndex(raftStatus)
+
+	// Can and should the raft logs be truncated?
+	truncatableIndexes := oldestIndex - firstIndex
```
up to here this looks like copy&paste from `shouldQueue`. Factor out the commonalities.

tamird commented:
Reviewed 4 of 4 files at r3.
Review status: all files reviewed at latest revision, 20 unresolved discussions.

---

<sup>**[storage/client_raft_log_queue_test.go, line 87 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2997#-K2Nwg08oMjXDuAw8gxb)** ([raw file](https://github.com/cockroachdb/cockroach/blob/2f07fdb6342c98c291f08f2d2260b0772b7af258/storage/client_raft_log_queue_test.go#L87)):</sup>
I would log the values here like you do above: `"raft log has not truncated, currentIndex:%d firstIndex:%d", currentIndex, firstIndex`
<!-- Reviewable comment -K2Nwg08oMjXDuAw8gxc:416268060 -->

---

<sup>**[storage/client_raft_log_queue_test.go, line 103 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2997#-K2Nwxf_8HfAGC3PsnDM)** ([raw file](https://github.com/cockroachdb/cockroach/blob/2f07fdb6342c98c291f08f2d2260b0772b7af258/storage/client_raft_log_queue_test.go#L103)):</sup>
Would be good to log the indexes here as well.
<!-- Reviewable comment -K2Nwxf_8HfAGC3PsnDN:-629915192 -->

---

<sup>**[storage/raft_log_queue.go, line 107 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2997#-K2NxkXCZoySgc_dnQQm)** ([raw file](https://github.com/cockroachdb/cockroach/blob/2f07fdb6342c98c291f08f2d2260b0772b7af258/storage/raft_log_queue.go#L107)):</sup>
"requires a raft log truncation" is ill-defined; best to define it in terms of some other function.
<!-- Reviewable comment -K2NxkXCZoySgc_dnQQn:-529575474 -->

---

<sup>**[storage/raft_log_queue.go, line 120 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2997#-K2NyCxwLnPSdFNDxJ7y)** ([raw file](https://github.com/cockroachdb/cockroach/blob/2f07fdb6342c98c291f08f2d2260b0772b7af258/storage/raft_log_queue.go#L120)):</sup>
I can't quite tell why `r.Desc()` is the right thing to mention here. Can you help me understand?
<!-- Reviewable comment -K2NyCxwLnPSdFNDxJ7z:1713375701 -->

---

<sup>**[storage/replica.go, line 1626 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2997#-K2NzFrGcvUyqcOdxueV)** ([raw file](https://github.com/cockroachdb/cockroach/blob/2f07fdb6342c98c291f08f2d2260b0772b7af258/storage/replica.go#L1626)):</sup>
From the depths of etcd/raft:
```
// Status returns the current status of the given group. Returns nil if no such group
// exists.
```
So let's make this error message more specific, and do the same in the test which deals with the same condition.
<!-- Reviewable comment -K2NzFrHUyhXnP1ym8ZB:322756322 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/2997)
<!-- Sent from Reviewable.io -->

BramGruneir commented:
Review status: 1 of 4 files reviewed at latest revision, 20 unresolved discussions.

---

<sup>**[storage/client_raft_log_queue_test.go, line 31 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2997#-K2DgvkEW0Hq3tashnk1-r3-31)** ([raw file](https://github.com/cockroachdb/cockroach/blob/2f07fdb6342c98c291f08f2d2260b0772b7af258/storage/client_raft_log_queue_test.go#L31)):</sup>
Oh, how did miss that.  Fixed.
<!-- Reviewable comment -K2NwnLuzZ7r10Z9h81x:808100832 -->

---

<sup>**[storage/client_raft_log_queue_test.go, line 31 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2997#-K2NvcMIYlB3lyCexipI)** ([raw file](https://github.com/cockroachdb/cockroach/blob/2f07fdb6342c98c291f08f2d2260b0772b7af258/storage/client_raft_log_queue_test.go#L31)):</sup>
Done.  I'm actually working up a test for that case right now.
<!-- Reviewable comment -K2NwumnOfj_F2E2UqTO:1772858704 -->

---

<sup>**[storage/client_raft_log_queue_test.go, line 39 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2997#-K2DgvkEW0Hq3tashnk1-r3-39)** ([raw file](https://github.com/cockroachdb/cockroach/blob/2f07fdb6342c98c291f08f2d2260b0772b7af258/storage/client_raft_log_queue_test.go#L39)):</sup>
Done.
<!-- Reviewable comment -K2NxI8EnWM5Sjs0_Qv4:-499146009 -->

---

<sup>**[storage/client_raft_log_queue_test.go, line 52 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2997#-K2DgvkEW0Hq3tashnk1-r3-52)** ([raw file](https://github.com/cockroachdb/cockroach/blob/2f07fdb6342c98c291f08f2d2260b0772b7af258/storage/client_raft_log_queue_test.go#L52)):</sup>
the test typically runs in... --- PASS: TestRaftLogQueue (0.01s)

I just wanted to future proof the test by using the Threshold value here.  Right now it's set to 1.  And I could see us maybe increasing it slightly for an extremely minor performance gain, but even then, it would only be to say 10 or so.
<!-- Reviewable comment -K2NxJiA9V9uZ0u1vEwU:1221448223 -->

---

<sup>**[storage/client_raft_log_queue_test.go, line 87 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2997#-K2Nwg08oMjXDuAw8gxb)** ([raw file](https://github.com/cockroachdb/cockroach/blob/2f07fdb6342c98c291f08f2d2260b0772b7af258/storage/client_raft_log_queue_test.go#L87)):</sup>
done
<!-- Reviewable comment -K2O0jtcyLIYBxUehxdM:677588487 -->

---

<sup>**[storage/client_raft_log_queue_test.go, line 102 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2997#-K2DgvkEW0Hq3tashnk1-r3-102)** ([raw file](https://github.com/cockroachdb/cockroach/blob/2f07fdb6342c98c291f08f2d2260b0772b7af258/storage/client_raft_log_queue_test.go#L102)):</sup>
yes, because we're looking only at the firstIndex, which will only get updated on another truncation.
<!-- Reviewable comment -K2O2VdSKS7QXD_qNcah:-161542370 -->

---

<sup>**[storage/client_raft_log_queue_test.go, line 103 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2997#-K2Nwxf_8HfAGC3PsnDM)** ([raw file](https://github.com/cockroachdb/cockroach/blob/2f07fdb6342c98c291f08f2d2260b0772b7af258/storage/client_raft_log_queue_test.go#L103)):</sup>
Done.
<!-- Reviewable comment -K2O0wEhOsoHodvpDIy6:-499146009 -->

---

<sup>**[storage/raft_log_queue.go, line 40 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2997#-K2DgvkEW0Hq3tashnk2-r3-40)** ([raw file](https://github.com/cockroachdb/cockroach/blob/2f07fdb6342c98c291f08f2d2260b0772b7af258/storage/raft_log_queue.go#L40)):</sup>
see my previous comment.
<!-- Reviewable comment -K2NymQIGJrgDlhXbm1u:-1206603018 -->

---

<sup>**[storage/raft_log_queue.go, line 71 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2997#-K2DgvkEW0Hq3tashnk2-r3-71)** ([raw file](https://github.com/cockroachdb/cockroach/blob/2f07fdb6342c98c291f08f2d2260b0772b7af258/storage/raft_log_queue.go#L71)):</sup>
Applied an even tighter bound, we definitely don't want to truncate it (or past it).
<!-- Reviewable comment -K2NzJTIWhZ_WrQSMUoJ:-1976763914 -->

---

<sup>**[storage/raft_log_queue.go, line 87 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2997#-K2NtPC-p3PSnmT0EJLH)** ([raw file](https://github.com/cockroachdb/cockroach/blob/2f07fdb6342c98c291f08f2d2260b0772b7af258/storage/raft_log_queue.go#L87)):</sup>
done
<!-- Reviewable comment -K2Nzacu3sXUAPtQggIT:677588487 -->

---

<sup>**[storage/raft_log_queue.go, line 88 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2997#-K2DgvkEW0Hq3tashnk2-r3-88)** ([raw file](https://github.com/cockroachdb/cockroach/blob/2f07fdb6342c98c291f08f2d2260b0772b7af258/storage/raft_log_queue.go#L88)):</sup>
There cannot be an error if raft leader is true.  But I'll make the change regardless.
<!-- Reviewable comment -K2Nz1GOFahOTEgUHmFQ:2027252475 -->

---

<sup>**[storage/raft_log_queue.go, line 103 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2997#-K2NrX79KWxKj139_S_C)** ([raw file](https://github.com/cockroachdb/cockroach/blob/2f07fdb6342c98c291f08f2d2260b0772b7af258/storage/raft_log_queue.go#L103)):</sup>
excellent point, I'll throw a warning if that's the case
<!-- Reviewable comment -K2Nzi7F-JG79sNx_bOz:1027773971 -->

---

<sup>**[storage/raft_log_queue.go, line 107 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2997#-K2NxkXCZoySgc_dnQQm)** ([raw file](https://github.com/cockroachdb/cockroach/blob/2f07fdb6342c98c291f08f2d2260b0772b7af258/storage/raft_log_queue.go#L107)):</sup>
cleaned
<!-- Reviewable comment -K2O1QLcMO-CdTkWoO4L:-1289388029 -->

---

<sup>**[storage/raft_log_queue.go, line 107 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2997#-K2DgvkEW0Hq3tashnk2-r3-107)** ([raw file](https://github.com/cockroachdb/cockroach/blob/2f07fdb6342c98c291f08f2d2260b0772b7af258/storage/raft_log_queue.go#L107)):</sup>
fixed
<!-- Reviewable comment -K2O1RKm513-z0-tyVVE:-467915377 -->

---

<sup>**[storage/raft_log_queue.go, line 111 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2997#-K2NtZ6KajaoA7xpUJR0)** ([raw file](https://github.com/cockroachdb/cockroach/blob/2f07fdb6342c98c291f08f2d2260b0772b7af258/storage/raft_log_queue.go#L111)):</sup>
Done.
<!-- Reviewable comment -K2O1TN-DTt9unVtg6-2:-499146009 -->

---

<sup>**[storage/raft_log_queue.go, line 112 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2997#-K2DgvkEW0Hq3tashnk2-r3-112)** ([raw file](https://github.com/cockroachdb/cockroach/blob/2f07fdb6342c98c291f08f2d2260b0772b7af258/storage/raft_log_queue.go#L112)):</sup>
Done.
<!-- Reviewable comment -K2O-5tc2FGCAe-WWmer:-499146009 -->

---

<sup>**[storage/raft_log_queue.go, line 126 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2997#-K2DgvkEW0Hq3tashnk2-r3-126)** ([raw file](https://github.com/cockroachdb/cockroach/blob/2f07fdb6342c98c291f08f2d2260b0772b7af258/storage/raft_log_queue.go#L126)):</sup>
Done.
<!-- Reviewable comment -K2O-6FRBEdLoWXMpdXO:-499146009 -->

---

<sup>**[storage/replica.go, line 1622 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2997#-K2Ntig-ZTJKZ0GYqcao)** ([raw file](https://github.com/cockroachdb/cockroach/blob/2f07fdb6342c98c291f08f2d2260b0772b7af258/storage/replica.go#L1622)):</sup>
changed this to just a quick function that returns the raft status.
<!-- Reviewable comment -K2O-GsKlJ6JSJEZ0rHu:1837462504 -->

---

<sup>**[storage/replica.go, line 1626 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2997#-K2NzFrGcvUyqcOdxueV)** ([raw file](https://github.com/cockroachdb/cockroach/blob/2f07fdb6342c98c291f08f2d2260b0772b7af258/storage/replica.go#L1626)):</sup>
Done.
<!-- Reviewable comment -K2O0DO25bQrDRoeN-in:-499146009 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/2997)
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 3 of 3 files at r4.
Review status: all files reviewed at latest revision, 24 unresolved discussions.

---

<sup>**[storage/client_raft_log_queue_test.go, line 65 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2997#-K2O5ohZRokn-LK890C-)** ([raw file](https://github.com/cockroachdb/cockroach/blob/d2b96fc71c17895ba12dce874a35771ad2964eb6/storage/client_raft_log_queue_test.go#L65)):</sup>
As mentioned elsewhere, this error really means that the raft group `desc.RangeID` doesn't exist.
<!-- Reviewable comment -K2O5oh_nQeDwIZVKtlz:1447524997 -->

---

<sup>**[storage/client_raft_log_queue_test.go, line 89 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2997#-K2O5xX7Ew2RJNw2nHtO)** ([raw file](https://github.com/cockroachdb/cockroach/blob/d2b96fc71c17895ba12dce874a35771ad2964eb6/storage/client_raft_log_queue_test.go#L89)):</sup>
missing "has"
<!-- Reviewable comment -K2O5xX8ZLz17sNzZCxK:-463516725 -->

---

<sup>**[storage/client_raft_log_queue_test.go, line 106 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2997#-K2O62PAFXUo19qtGYxQ)** ([raw file](https://github.com/cockroachdb/cockroach/blob/d2b96fc71c17895ba12dce874a35771ad2964eb6/storage/client_raft_log_queue_test.go#L106)):</sup>
s/original/first/
<!-- Reviewable comment -K2O62PAFXUo19qtGYxR:1614006652 -->

---

<sup>**[storage/raft_log_queue.go, line 103 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2997#-K2NrX79KWxKj139_S_C)** ([raw file](https://github.com/cockroachdb/cockroach/blob/2f07fdb6342c98c291f08f2d2260b0772b7af258/storage/raft_log_queue.go#L103)):</sup>
alternatively, could you change the below to `return oldestIndex > firstIndex + RaftLogQueueLogSizeThreshold`?
<!-- Reviewable comment -K2O6dxPwmR14ay8okAs:-1221133117 -->

---

<sup>**[storage/raft_log_queue.go, line 88 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2997#-K2O6C7AG7_u7xOlZDuH)** ([raw file](https://github.com/cockroachdb/cockroach/blob/d2b96fc71c17895ba12dce874a35771ad2964eb6/storage/raft_log_queue.go#L88)):</sup>
as mentioned elsewhere, this means the group doesn't exist, and in this case, should probably be more than a warning?
<!-- Reviewable comment -K2O6C7AG7_u7xOlZDuI:1936135012 -->

---

<sup>**[storage/raft_log_queue.go, line 114 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2997#-K2O70YyJDWj3qZm3qvO)** ([raw file](https://github.com/cockroachdb/cockroach/blob/d2b96fc71c17895ba12dce874a35771ad2964eb6/storage/raft_log_queue.go#L114)):</sup>
i think the word "old" is not useful here
also clarify that "size" is just about quantity; it's currently unclear.
<!-- Reviewable comment -K2O70YyJDWj3qZm3qvP:536312912 -->

---

<sup>**[storage/raft_log_queue.go, line 118 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2997#-K2O7ClUuE-tq4kjgNqe)** ([raw file](https://github.com/cockroachdb/cockroach/blob/d2b96fc71c17895ba12dce874a35771ad2964eb6/storage/raft_log_queue.go#L118)):</sup>
again, means the group doesn't exist.
<!-- Reviewable comment -K2O7ClVF8mBXPu6Ob_C:545933190 -->

---

<sup>**[storage/replica.go, line 1622 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2997#-K2O7aKoDs_3bWyyoVse)** ([raw file](https://github.com/cockroachdb/cockroach/blob/d2b96fc71c17895ba12dce874a35771ad2964eb6/storage/replica.go#L1622)):</sup>
something messed up about this comment.
<!-- Reviewable comment -K2O7aKpW9OVDnDwF5So:-1995050889 -->

---

<sup>**[storage/replica.go, line 1623 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2997#-K2O7hnU3U3qFUXeoGsf)** ([raw file](https://github.com/cockroachdb/cockroach/blob/d2b96fc71c17895ba12dce874a35771ad2964eb6/storage/replica.go#L1623)):</sup>
This function has two callers, but those are the parts that are identical between shouldQueue and process; once you factor out those commonalities, think function should be removed (inlined).
<!-- Reviewable comment -K2O7hnU3U3qFUXeoGsg:1888547395 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/2997)
<!-- Sent from Reviewable.io -->

BramGruneir commented:
Review status: 1 of 3 files reviewed at latest revision, 24 unresolved discussions.

---

<sup>**[storage/client_raft_log_queue_test.go, line 65 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2997#-K2O5ohZRokn-LK890C-)** ([raw file](https://github.com/cockroachdb/cockroach/blob/d2b96fc71c17895ba12dce874a35771ad2964eb6/storage/client_raft_log_queue_test.go#L65)):</sup>
Yes,, but they're created lazily, which in this case would be a fatal error but in other cases should be fine.  Also re-worded.
<!-- Reviewable comment -K2OA-GO9FwS-4B3OP_P:-506883358 -->

---

<sup>**[storage/client_raft_log_queue_test.go, line 89 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2997#-K2O5xX7Ew2RJNw2nHtO)** ([raw file](https://github.com/cockroachdb/cockroach/blob/d2b96fc71c17895ba12dce874a35771ad2964eb6/storage/client_raft_log_queue_test.go#L89)):</sup>
Done.
<!-- Reviewable comment -K2OAIZLigIrAZmCxKHI:-499146009 -->

---

<sup>**[storage/client_raft_log_queue_test.go, line 106 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2997#-K2O62PAFXUo19qtGYxQ)** ([raw file](https://github.com/cockroachdb/cockroach/blob/d2b96fc71c17895ba12dce874a35771ad2964eb6/storage/client_raft_log_queue_test.go#L106)):</sup>
actually, finalIndex, but re-worded and fixed
<!-- Reviewable comment -K2OAOOy-Mt9-41xXlKf:-1974381845 -->

---

<sup>**[storage/raft_log_queue.go, line 103 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2997#-K2NrX79KWxKj139_S_C)** ([raw file](https://github.com/cockroachdb/cockroach/blob/2f07fdb6342c98c291f08f2d2260b0772b7af258/storage/raft_log_queue.go#L103)):</sup>
changed this all up a bit by grouping most of the common code between shouldQueue and process
<!-- Reviewable comment -K2OAvmJbMPrLs_zKuYH:-753609928 -->

---

<sup>**[storage/raft_log_queue.go, line 120 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2997#-K2NyCxwLnPSdFNDxJ7y)** ([raw file](https://github.com/cockroachdb/cockroach/blob/2f07fdb6342c98c291f08f2d2260b0772b7af258/storage/raft_log_queue.go#L120)):</sup>
Changed this to only printout the range id
<!-- Reviewable comment -K2OC1cNLoNdpQtzn9YH:-1164358900 -->

---

<sup>**[storage/raft_log_queue.go, line 88 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2997#-K2O6C7AG7_u7xOlZDuH)** ([raw file](https://github.com/cockroachdb/cockroach/blob/d2b96fc71c17895ba12dce874a35771ad2964eb6/storage/raft_log_queue.go#L88)):</sup>
this is possible if we just haven't had any operations on the range yet.  Possible, but only right at the start.
<!-- Reviewable comment -K2OAiCZvq6sirZ1SfKW:-1284786498 -->

---

<sup>**[storage/raft_log_queue.go, line 114 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2997#-K2O70YyJDWj3qZm3qvO)** ([raw file](https://github.com/cockroachdb/cockroach/blob/d2b96fc71c17895ba12dce874a35771ad2964eb6/storage/raft_log_queue.go#L114)):</sup>
reworded to clarify
<!-- Reviewable comment -K2OBhS_hTdr59wDE8Fm:1566593564 -->

---

<sup>**[storage/raft_log_queue.go, line 118 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2997#-K2O7ClUuE-tq4kjgNqe)** ([raw file](https://github.com/cockroachdb/cockroach/blob/d2b96fc71c17895ba12dce874a35771ad2964eb6/storage/raft_log_queue.go#L118)):</sup>
Ok, I've made it even more explicit
<!-- Reviewable comment -K2OBkC7gEMYwl2700T5:603131041 -->

---

<sup>**[storage/replica.go, line 1622 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2997#-K2O7aKoDs_3bWyyoVse)** ([raw file](https://github.com/cockroachdb/cockroach/blob/d2b96fc71c17895ba12dce874a35771ad2964eb6/storage/replica.go#L1622)):</sup>
reworded.. well, removed actually
<!-- Reviewable comment -K2OCjpt5nM3vmtK9rm_:-433401220 -->

---

<sup>**[storage/replica.go, line 1623 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2997#-K2O7hnU3U3qFUXeoGsf)** ([raw file](https://github.com/cockroachdb/cockroach/blob/d2b96fc71c17895ba12dce874a35771ad2964eb6/storage/replica.go#L1623)):</sup>
removed
<!-- Reviewable comment -K2OCoW8a6ro6vCuNmeC:-1054325829 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/2997)
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 3 of 3 files at r5.
Review status: all files reviewed at latest revision, 24 unresolved discussions, some commit checks failed.

---

<sup>**[storage/client_raft_log_queue_test.go, line 65 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2997#-K2O5ohZRokn-LK890C-)** ([raw file](https://github.com/cockroachdb/cockroach/blob/d2b96fc71c17895ba12dce874a35771ad2964eb6/storage/client_raft_log_queue_test.go#L65)):</sup>
Why is it relevant that they're created lazily? What I'm saying is that you can make this error more informative by saying "the raft group doesn't exist", rather than the current error (even after your edit). It seems like the RangeID is the only interesting thing to log here, rather than the entire descriptor.
<!-- Reviewable comment -K2OHTCt1Sjrtg4Jjodd:-727661050 -->

---

<sup>**[storage/client_raft_log_queue_test.go, line 106 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2997#-K2O62PAFXUo19qtGYxQ)** ([raw file](https://github.com/cockroachdb/cockroach/blob/d2b96fc71c17895ba12dce874a35771ad2964eb6/storage/client_raft_log_queue_test.go#L106)):</sup>
I find this new error message more confusing; what is "actual"?
<!-- Reviewable comment -K2OI8AHTWsIdpuiL6sT:82852273 -->

---

<sup>**[storage/client_raft_log_queue_test.go, line 89 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2997#-K2OHr3jp-WeMXb9UxB5)** ([raw file](https://github.com/cockroachdb/cockroach/blob/31c7369ecc85df334ba0b86b7c88491eac47af38/storage/client_raft_log_queue_test.go#L89)):</sup>
the words in this error do not line up with the variables being logged.
<!-- Reviewable comment -K2OHr3kz-g61zDZduNa:-648783062 -->

---

<sup>**[storage/raft_log_queue.go, line 103 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2997#-K2NrX79KWxKj139_S_C)** ([raw file](https://github.com/cockroachdb/cockroach/blob/2f07fdb6342c98c291f08f2d2260b0772b7af258/storage/raft_log_queue.go#L103)):</sup>
Seems like you're doing the same unsafe arithmetic in the new place.
<!-- Reviewable comment -K2OJ4J47VBRg7EaxN0x:1598159259 -->

---

<sup>**[storage/raft_log_queue.go, line 114 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2997#-K2O70YyJDWj3qZm3qvO)** ([raw file](https://github.com/cockroachdb/cockroach/blob/d2b96fc71c17895ba12dce874a35771ad2964eb6/storage/raft_log_queue.go#L114)):</sup>
I still don't understand what "stale" means.
<!-- Reviewable comment -K2OJKRiO_QjN-5oPdHH:1928033139 -->

---

<sup>**[storage/raft_log_queue.go, line 67 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2997#-K2OIk4ac6xo8mCCtZ4a)** ([raw file](https://github.com/cockroachdb/cockroach/blob/31c7369ecc85df334ba0b86b7c88491eac47af38/storage/raft_log_queue.go#L67)):</sup>
what does "old" mean?
<!-- Reviewable comment -K2OIk4ac6xo8mCCtZ4b:63144967 -->

---

<sup>**[storage/raft_log_queue.go, line 73 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2997#-K2OItQgSJhcveIiXxM7)** ([raw file](https://github.com/cockroachdb/cockroach/blob/31c7369ecc85df334ba0b86b7c88491eac47af38/storage/raft_log_queue.go#L73)):</sup>
s/%s/%d/

As mentioned elsewhere, this error message can be made more informative.
<!-- Reviewable comment -K2OItQgSJhcveIiXxM8:568371929 -->

---

<sup>**[storage/raft_log_queue.go, line 91 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2997#-K2OIS9V1jzNQOhaXZrm)** ([raw file](https://github.com/cockroachdb/cockroach/blob/31c7369ecc85df334ba0b86b7c88491eac47af38/storage/raft_log_queue.go#L91)):</sup>
s/range: %s/range: %d/
<!-- Reviewable comment -K2OIS9W1qC4PTR2ieWV:-1397780980 -->

---

<sup>**[storage/raft_log_queue.go, line 99 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2997#-K2OIbfmTYRr6JQQKCNx)** ([raw file](https://github.com/cockroachdb/cockroach/blob/31c7369ecc85df334ba0b86b7c88491eac47af38/storage/raft_log_queue.go#L99)):</sup>
is this safe? unsigned arithmetic and so forth.
<!-- Reviewable comment -K2OIbfmTYRr6JQQKCNy:-1584040542 -->

---

<sup>**[storage/raft_log_queue.go, line 128 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2997#-K2OJUPqyRwXfJg1qyxC)** ([raw file](https://github.com/cockroachdb/cockroach/blob/31c7369ecc85df334ba0b86b7c88491eac47af38/storage/raft_log_queue.go#L128)):</sup>
This constant's name is a mouthful, but I think this rename only makes it worse.
<!-- Reviewable comment -K2OJUPqyRwXfJg1qyxD:-792129463 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/2997)
<!-- Sent from Reviewable.io -->

BramGruneir commented:
Review status: 1 of 3 files reviewed at latest revision, 24 unresolved discussions.

---

<sup>**[storage/client_raft_log_queue_test.go, line 65 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2997#-K2O5ohZRokn-LK890C-)** ([raw file](https://github.com/cockroachdb/cockroach/blob/d2b96fc71c17895ba12dce874a35771ad2964eb6/storage/client_raft_log_queue_test.go#L65)):</sup>
Changed it to the range ID.  But I don't know why this is showing an older version of the error message.  It should say "raft status does not exist for range %s".  But I'll update to your wording.
<!-- Reviewable comment -K2OOvqZ8XVJOcoCBYYO:-123009375 -->

---

<sup>**[storage/client_raft_log_queue_test.go, line 106 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2997#-K2O62PAFXUo19qtGYxQ)** ([raw file](https://github.com/cockroachdb/cockroach/blob/d2b96fc71c17895ba12dce874a35771ad2964eb6/storage/client_raft_log_queue_test.go#L106)):</sup>
fixed
<!-- Reviewable comment -K2OQUcpCcYKGQ8CaN_o:-467915377 -->

---

<sup>**[storage/client_raft_log_queue_test.go, line 89 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2997#-K2OHr3jp-WeMXb9UxB5)** ([raw file](https://github.com/cockroachdb/cockroach/blob/31c7369ecc85df334ba0b86b7c88491eac47af38/storage/client_raft_log_queue_test.go#L89)):</sup>
renamed the variable names
<!-- Reviewable comment -K2OPf16ObOoQpK3QiMX:-526405006 -->

---

<sup>**[storage/raft_log_queue.go, line 103 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2997#-K2NrX79KWxKj139_S_C)** ([raw file](https://github.com/cockroachdb/cockroach/blob/2f07fdb6342c98c291f08f2d2260b0772b7af258/storage/raft_log_queue.go#L103)):</sup>
Done.
<!-- Reviewable comment -K2OS3N9oeBACLoKOlfy:-499146009 -->

---

<sup>**[storage/raft_log_queue.go, line 114 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2997#-K2O70YyJDWj3qZm3qvO)** ([raw file](https://github.com/cockroachdb/cockroach/blob/d2b96fc71c17895ba12dce874a35771ad2964eb6/storage/raft_log_queue.go#L114)):</sup>
added a defn for stale under the constant
<!-- Reviewable comment -K2OTOGOZQvvkt0Q85be:-1145941172 -->

---

<sup>**[storage/raft_log_queue.go, line 67 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2997#-K2OIk4ac6xo8mCCtZ4a)** ([raw file](https://github.com/cockroachdb/cockroach/blob/31c7369ecc85df334ba0b86b7c88491eac47af38/storage/raft_log_queue.go#L67)):</sup>
renamed the function
<!-- Reviewable comment -K2ORaYofPqDy6ezqdsb:-614127578 -->

---

<sup>**[storage/raft_log_queue.go, line 73 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2997#-K2OItQgSJhcveIiXxM7)** ([raw file](https://github.com/cockroachdb/cockroach/blob/31c7369ecc85df334ba0b86b7c88491eac47af38/storage/raft_log_queue.go#L73)):</sup>
Done.
<!-- Reviewable comment -K2ORlofxO47Re8vnrIi:-499146009 -->

---

<sup>**[storage/raft_log_queue.go, line 91 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2997#-K2OIS9V1jzNQOhaXZrm)** ([raw file](https://github.com/cockroachdb/cockroach/blob/31c7369ecc85df334ba0b86b7c88491eac47af38/storage/raft_log_queue.go#L91)):</sup>
Done.
<!-- Reviewable comment -K2ORoNd14VNG78z76DZ:-499146009 -->

---

<sup>**[storage/raft_log_queue.go, line 99 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2997#-K2OIbfmTYRr6JQQKCNx)** ([raw file](https://github.com/cockroachdb/cockroach/blob/31c7369ecc85df334ba0b86b7c88491eac47af38/storage/raft_log_queue.go#L99)):</sup>
Done.
<!-- Reviewable comment -K2OS38TvrLetvK90L_y:-499146009 -->

---

<sup>**[storage/raft_log_queue.go, line 128 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2997#-K2OJUPqyRwXfJg1qyxC)** ([raw file](https://github.com/cockroachdb/cockroach/blob/31c7369ecc85df334ba0b86b7c88491eac47af38/storage/raft_log_queue.go#L128)):</sup>
renamed to RaftLogQueueStaleThreshold and re-worded the comment around the constant.
<!-- Reviewable comment -K2OSv5a1mdaHGuSxUQ_:-1772049284 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/2997)
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 2 of 2 files at r6.
Review status: all files reviewed at latest revision, 18 unresolved discussions.

---

<sup>**[storage/client_raft_log_queue_test.go, line 65 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2997#-K2O5ohZRokn-LK890C-)** ([raw file](https://github.com/cockroachdb/cockroach/blob/d2b96fc71c17895ba12dce874a35771ad2964eb6/storage/client_raft_log_queue_test.go#L65)):</sup>
"raft status does not exist for range %s" was what I saw.
<!-- Reviewable comment -K2OV2cqN0dAiBVai2Yf:-267436602 -->

---

<sup>**[storage/raft_log_queue.go, line 67 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2997#-K2OIk4ac6xo8mCCtZ4a)** ([raw file](https://github.com/cockroachdb/cockroach/blob/31c7369ecc85df334ba0b86b7c88491eac47af38/storage/raft_log_queue.go#L67)):</sup>
wha? what does "old" in this comment mean?
<!-- Reviewable comment -K2OVCSYMqHQWR5q6dzk:-104143432 -->

---

<sup>**[storage/raft_log_queue.go, line 39 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2997#-K2OVp_uhxSGVyFrf8W3)** ([raw file](https://github.com/cockroachdb/cockroach/blob/7fcc8a8cd8b60ced5f69038c4132153cafc19e77/storage/raft_log_queue.go#L39)):</sup>
s/in which/ which/, s/in the range/of the range/
<!-- Reviewable comment -K2OVp_vnMPaoJlvI_Pi:1560135076 -->

---

<sup>**[storage/raft_log_queue.go, line 40 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2997#-K2OWViiksVWOSSW4fHs)** ([raw file](https://github.com/cockroachdb/cockroach/blob/7fcc8a8cd8b60ced5f69038c4132153cafc19e77/storage/raft_log_queue.go#L40)):</sup>
Not sure about this, but s/advanced past/acknowledged/. I don't remember what Progress.Match really corresponds to, but we should use the raft terminology here.
<!-- Reviewable comment -K2OWViiksVWOSSW4fHt:-1757711003 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/2997)
<!-- Sent from Reviewable.io -->

BramGruneir commented:
Review status: all files reviewed at latest revision, 19 unresolved discussions.

---

<sup>**[storage/client_raft_log_queue_test.go, line 65 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2997#-K2O5ohZRokn-LK890C-)** ([raw file](https://github.com/cockroachdb/cockroach/blob/d2b96fc71c17895ba12dce874a35771ad2964eb6/storage/client_raft_log_queue_test.go#L65)):</sup>
Done.
<!-- Reviewable comment -K2OXNdBy-y8j6VKIDzt:-499146009 -->

---

<sup>**[storage/raft_log_queue.go, line 67 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2997#-K2OIk4ac6xo8mCCtZ4a)** ([raw file](https://github.com/cockroachdb/cockroach/blob/31c7369ecc85df334ba0b86b7c88491eac47af38/storage/raft_log_queue.go#L67)):</sup>
renamed to stale
<!-- Reviewable comment -K2OXEy23BUIlrq9vS_i:-729400393 -->

---

<sup>**[storage/raft_log_queue.go, line 39 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2997#-K2OVp_uhxSGVyFrf8W3)** ([raw file](https://github.com/cockroachdb/cockroach/blob/7fcc8a8cd8b60ced5f69038c4132153cafc19e77/storage/raft_log_queue.go#L39)):</sup>
Done.
<!-- Reviewable comment -K2OXa3We6tA0ukoQ3pJ:-499146009 -->

---

<sup>**[storage/raft_log_queue.go, line 40 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2997#-K2OWViiksVWOSSW4fHs)** ([raw file](https://github.com/cockroachdb/cockroach/blob/7fcc8a8cd8b60ced5f69038c4132153cafc19e77/storage/raft_log_queue.go#L40)):</sup>
seems to be "progressed" past.  Hence it being under 
<!-- Reviewable comment -K2OYElc8uO_b92qK0xk:212379581 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/2997)
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 1 files at r7.
Review status: all files reviewed at latest revision, 16 unresolved discussions.

---

<sup>**[storage/raft_log_queue.go, line 39 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2997#-K2OacQIKSF7qs0J6iWi)** ([raw file](https://github.com/cockroachdb/cockroach/blob/ddddc63945c24392bb30c9eb59a28ed608d8fa9c/storage/raft_log_queue.go#L39)):</sup>
s/ for//
<!-- Reviewable comment -K2OacQIKSF7qs0J6iWj:524935530 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/2997)
<!-- Sent from Reviewable.io -->

BramGruneir commented:
Review status: 2 of 3 files reviewed at latest revision, 16 unresolved discussions.

---

<sup>**[storage/raft_log_queue.go, line 39 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/2997#-K2OacQIKSF7qs0J6iWi)** ([raw file](https://github.com/cockroachdb/cockroach/blob/ddddc63945c24392bb30c9eb59a28ed608d8fa9c/storage/raft_log_queue.go#L39)):</sup>
Done.
<!-- Reviewable comment -K2Od8MJvzSNRtOEAvJY:-499146009 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/2997)
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 1 files at r8.
Review status: all files reviewed at latest revision, 15 unresolved discussions.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/2997)
<!-- Sent from Reviewable.io -->
